lib/srv/db/cloud/users/user.go

```diff
@@ -0,0 +1,215 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package users
+
+import (
+	"context"
+	"time"
+
+	"github.com/gravitational/teleport/lib/secrets"
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"github.com/sirupsen/logrus"
+)
+
+// User represents a managed cloud database user.
+type User interface {
+	// GetID returns a globally unique ID for the user.
+	GetID() string
+	// GetInDatabaseName returns in-database username for the user.
+	GetInDatabaseName() string
+	// Setup preforms any setup necessary like creating password secret.
+	Setup(ctx context.Context) error
+	// Teardown performs any teardown necessary like deleting password secret.
+	Teardown(ctx context.Context) error
+	// GetPassword returns the password used for database login.
+	GetPassword(ctx context.Context) (string, error)
+	// RotatePassword rotates user's password.
+	RotatePassword(ctx context.Context) error
+}
+
+// baseUser is a base implementation of User.
+type baseUser struct {
```

Comment:
What do you think  about renaming the `baseUse`  to  `ElasticCacheUser` ?

Reply:
I would like to keep `baseUser` as it is used by MemoryDB as well. The only difference is it passes a different `modifyUserFunc` callback. I was trying to make it generic as it only has secret manager logic. Hopefully, it's good for other database users in the future or maybe we can refactor when that happens.